rfcs/2020-08-31-3645-graphql-api.md

```diff
@@ -0,0 +1,279 @@
+
+I anticipate doing something similar with the Vector API. async-graphql has a [Context](https://async-graphql.github.io/async-graphql/en/context.html) struct, which is typically used for passing in shared resources such as a database connection pool, or request-specific data such as the current user session.
+
+This is largely TBD, but the basic mechanisms are there to allow for flexible auth when we need it.
+
+## Doc-level Proposal
+
+A Vector observability layer has already been agreed internally. Work is underway.
+
+This proposal discusses the protocol which will govern communications between a running Vector instance, and a web UI and CLI.
+
+The result of this proposal won't directly impact user interaction with observability tooling.
+
+## Rationale
+
+We have built a decent body of experience with GraphQL at Timber, albeit on non-Vector projects.
+
+The tooling I am proposing here represents the same stack (save for swapping Apollo for Urql.)
+
+In a previous Timber project, we used [gqlgen](https://github.com/99designs/gqlgen) as our server library, written in Go. The tooling with async-graphql is similar, though Rust's language features enable a more composible 'code first' approach using macros.
+
+I believe the type of data we are consuming benefits from a strongly typed interface, and that compile-time client generation will significantly reduce the time-to-market of what will be already be a very complex front-end web app.
+
+## Prior Art
+
+See above
+
+## Drawbacks
+
+* GraphQL is generally geared toward graph data (hence the name), although this isn't a hard requirement. IMO, it's a solid general-purpose API interface that mimics what we'd generally wind up creating internally anyway.
+* Queries with unknown levels of nesting can't be queried statically; generally, a need for queries with flatter structures arise.
+* Versioning is trickier, given the lack of distinct endpoints. Nothing stopping multiple endpoints representing versions, although typically this is done in the schema.
+* Some annoying rough edges -- the lack of input unions, for example, typically worked around with more specific queries.
+* Reliance on a relatively young library rather than a 'pure' client like a stdlib tcp server.
+* A need to invest, longer-term, in understanding and contributing to an external library.
+
+## Alternatives
+
+Essentially any browser-compatible protocol could be used, and any text/binary format. This could include other libs such as Protobuf, which might form of a hybrid of the aforementioned approaches.
+
+The maturity of tooling would need further investigation, as well as the development experience of working with any given format.
+
+## Outstanding Questions
+
+* What are the performance characteristics we care about?
```

Comment:
I'm most curious about data streaming performance. Subscriptions seem like a reasonable abstraction, but I haven't dug into how they're implemented what kind of performance we should be able to expect.

Again, this is probably fine for the browser-based UI but if we're going to commit to a general protocol for getting data out of vector, I'd like to make sure it's going to hold up to other use cases. For example, our current `vector` source and sink use protocol buffers over TCP. It'd be very cool if we could replace that with something more general that allows filtering streams, etc, but I imagine that falls pretty far outside of the normal GraphQL use case. What kinds of issues might we run into if we tried to go down that path? Could we swap out WebSockets for normal sockets when going backend-to-backend?

Reply:
> I'm most curious about data streaming performance. Subscriptions seem like a reasonable abstraction, but I haven't dug into how they're implemented what kind of performance we should be able to expect.

async-graphql adds very little to subscriptions; it's mostly Warp (in our implementation.)

There's some bytes -> utf8 copying, which could perhaps be improved:

```rust
select! {
    bytes = srx.next() => {
        if let Some(bytes) = bytes {
            if let Ok(text) = String::from_utf8(bytes.to_vec()) {
                if tx.send(Message::text(text)).await.is_err()
                {
                    return;
                }
            }
        } else {
            return;
        }
    }
    msg = rx.next() => {
        if let Some(Ok(msg)) = msg {
            if msg.is_text() {
                if stx.send(Bytes::copy_from_slice(msg.as_bytes())).await.is_err() {
                    return;
                }
            }
        } else {
            return;
        }
    }
}
```

The `tx` channel is the default futures-util `SplitSink` that Warp provides on WS upgraded connections. Aside from iterating over a stream and sending it down the pipe, there's no real magic.

If warp is [too slow](https://www.techempower.com/benchmarks/), there's an [actix-web](https://crates.io/crates/async-graphql-actix-web) integration. Or we could create a TCP wrapper of our own.

Based on [Hashrocket's WS shoot-out](https://github.com/hashrocket/websocket-shootout/blob/master/results/round-02.md#rust---ws) (and all things being equal), we'd likely have to exceed 5,000 - 10,000+ connected clients for WS transmission to bust the 100ms mark. Given we're targeting single users (or at most, a handful) and metrics that currently tick over at 2 second intervals... even with dozens of multi-KB payloads multiplexed over the same connection, the constraint is likely to be network I/O before it's anything instrinsically non-performant about WebSockets.

This is completely unscientific, and running on a relatively beefy 8-core i9 laptop with 32gb RAM, but here's a quick screencast showing 9 connected Chrome clients observing a single Vector instance, with events generated linearly at at 5ms intervals, aggregrated at 100ms intervals (per client), with all the overhead of 9x Chrome windows running 9x v8 stacks with 9x React and 9x urql clients on 9x local Websocket connections...

https://www.loom.com/share/b94f952fd520491bb2bec30a8dd07c34

Pause the video on any frame. Event numbers are exactly in sync, +/- 99ms to account for variations between intervals.

> Again, this is probably fine for the browser-based UI but if we're going to commit to a general protocol for getting data out of vector, I'd like to make sure it's going to hold up to other use cases. 

GraphQL is ideally suited to a general API, because it's just typed JSON over HTTP. Every language has a stdlib or de facto HTTP client that would work with it. This applies to every query/mutation (i.e. GET/POST-type query)

Streaming data is always going to require a bit more thought, by virtue of its opinionated protocols and inherently longer-lived connections. I don't think the choice between GraphQL or something has necessarily has to be binary. If we want to offer an alternative protocol, I see at least two options:

1. Use the GraphQL spec. Like the Warp WS snippet, this is just a case of iterating over a stream and sending it down whatever channel backs into the connected client. Same I/O otherwise.

2. Ignore the GraphQL spec. gRPC might be a better choice for some clients (IMO it's a poor choice for the web, for the reasons outlined in the RFC, but probably preferable for service-to-service connections.) In which case, the gRPC would drive the interface for that particular gateway, and likely implement its own abstraction over our internal types, auth, and whatever else is specific to gRPC.

As a 'general purpose' API though... one that has the least baggage from a simple protocol to handle both one-time queries and streaming data... plain ol' HTTP and WebSockets with JSON payloads is probably as simple and therefore as universally applicable as any other choice we could make.

Reply:
If necessary, I will improve.

Reply:
I think @leebenson did a great job explaining above.  I just want to amplify one piece that he mentioned in passing: multiplexing.  GraphQL subscriptions are meant to be light weight.  In a UI backed by GraphQL they can be created and destroyed at will as the user is moving around in the app (when FB introduced subscriptions, they explained that they were using these to subscribe and unsubscribe users to "likes" on every piece of content as they were scrolling through their feed).  GraphQL subscriptions are a way to multiplex lots of messages over a single connection.  To me, this goes right at the the performance question.  This paradigm provides a way for clients to subscribe and unsubscribe over a single multiplexed connection to exactly the content they want with an expressive query language with dramatically less over-fetching.  The single connection also means there is not all the setup and tear down cost of an HTTP request for each of these interactions.

Reply:
>  It'd be very cool if we could replace that with something more general that allows filtering streams, etc, but I imagine that falls pretty far outside of the normal GraphQL use case.

If you mean 'take a single stream, and return from it only what the client asked for', then that's very much in GraphQL's wheelhouse and a common use-case for subscriptions.

There are two ways we can leverage GraphQL for this:

### Define specific queries

Instead of a catch-all `metrics` subscription, we could have multiple `subscription` definitions which filter on the same underlying stream, but only return the relevant type.

For example -- imagining a 'bytes' and an 'events' metric, the queries could look like:

```gql
{
  bytesMetrics {
    bytesProcessed
  }
}
```

and

```gql
subscription {
  eventsMetrics {
    numberOfEvents
  }
}
```

An example of this approach is in #3514. I have a func that returns a 'global' controller, like this...

```rust
/// Returns a stream of `Metric`s, collected at the provided millisecond interval
fn get_metrics(interval: i32) -> impl Stream<Item = Metric> {
    let controller = get_controller().unwrap();
    let mut interval = tokio::time::interval(Duration::from_millis(interval as u64));

    stream! {
        while let _ = interval.tick().await {
            for ev in capture_metrics(&controller) {
                if let Event::Metric(m) = ev {
                    yield m
                }
            }
        }
    }
}
```

... and a `metrics` subscription which (currently) only returns 'events_processed' metrics, like this:

```rust
#[Subscription]
impl MetricsSubscription {
    /// Returns all Vector metrics, aggregated at the provided millisecond interval
    async fn metrics(
        &self,
        #[arg(validator(IntRange(min = "100", max = "60_000")))] interval: i32,
    ) -> impl Stream<Item = MetricType> {
        get_metrics(interval).filter_map(|m| match m.name.as_str() {
            "events_processed" => Some(MetricType::EventsProcessed(m.into())),
            _ => None,
        })
    }
}
```

For additional metrics types, we'd just create other funcs representing the parts of the stream the user firing that subscription cares about. Everything else would get mapped to `None`, and therefore would be ignored.

### Using parameters

We could also use input params to control the stream. So by default, a stream would return everything -- but could be filtered to 'subscribe' to just the streams it cares about.

Pseudo:

```rust
enum SubscribeToMetric {
    Events,
    Bytes,
}

#[Subscription]
impl MetricsSubscription {
    /// Returns all Vector metrics, aggregated at the provided millisecond interval
    async fn metrics(
        &self,
        #[arg(validator(IntRange(min = "100", max = "60_000")))] interval: i32,
        subscribe_to: SubscribeToMetric,
    ) -> impl Stream<Item = MetricType> {
        get_metrics(interval).filter_map(|m| match (m.name.as_str(), subscribe_to) {
            ("events_processed", SubscribeToMetric::Events) => {
                Some(MetricType::EventsProcessed(m.into()))
            }
            ("bytes_processed", &SubscribeToMetric::Bytes) => {
                Some(MetricType::EventsProcessed(m.into()))
            }
            _ => None,
        })
    }
}
```

In this case, both the events and the params would need to match, for the metric to be propagated to the WebSocket client.

---

This doesn't have to be limited to a single metric. We could take an array of metric types, and return them as part of the same subscription. Or break them all up, and let the client multiplex per @sghall's comments. A combination of well-defined subscriptions and flexible input filters could yield whatever we need.

Reply:
Haha, you found an unnecessary memory copy, I optimized it, and also optimized some other places about subscriptions. 😁

Reply:
Thanks @sunli829, that's awesome!

Reply:
Makes sense. As a backend person, WebSockets definitely stick out as an "extra" thing I would never normally use outside of the browser, but the difference in overhead between them and normal TCP sockets is likely not high.

> We could also use input params to control the stream. So by default, a stream would return everything -- but could be filtered to 'subscribe' to just the streams it cares about.

This would definitely be an interesting route. Allowing a client to fetch a catalog of the data that's available and then send a "query" that returns of stream of just the data matching that query would be really cool.